src/test/storage/dictionary_segment_test.cpp

```diff
@@ -0,0 +1,161 @@
+  EXPECT_EQ(dict_segment->get_typed_value(3), "Steve");
+  EXPECT_EQ(dict_segment->get_typed_value(6), std::nullopt);
+
+#ifdef DEBUG
+  EXPECT_THROW(dict_segment->get(6), std::logic_error);
+  EXPECT_THROW(dict_segment->get(7), std::logic_error);
+#endif
+
+  EXPECT_EQ(dict_segment->value_of_value_id(ValueID{0}), "Alexander");
+
+  EXPECT_EQ(dict_segment->upper_bound(NULL_VALUE), INVALID_VALUE_ID);
+  EXPECT_EQ(dict_segment->lower_bound(NULL_VALUE), INVALID_VALUE_ID);
+}
+
+TEST_F(StorageDictionarySegmentTest, NullValues) {
+  auto null_values_count = u_int8_t{4};
+
+  for (auto counter = u_int8_t{0}; counter < null_values_count; ++counter) {
+    value_segment_str->append(NULL_VALUE);
+  }
+
+  const auto dict_segment = std::make_shared<DictionarySegment<std::string>>(value_segment_str);
+
+  EXPECT_EQ(dict_segment->size(), 4);
+
+  EXPECT_EQ(dict_segment->unique_values_count(), 0);
+
+  for (auto index = ChunkOffset{0}; index < null_values_count; ++index) {
+    EXPECT_EQ(dict_segment->attribute_vector()->get(index), dict_segment->null_value_id());
+    EXPECT_EQ(dict_segment->get_typed_value(index), std::nullopt);
+  }
+
+  EXPECT_EQ(dict_segment->upper_bound(NULL_VALUE), INVALID_VALUE_ID);
+  EXPECT_EQ(dict_segment->lower_bound(NULL_VALUE), INVALID_VALUE_ID);
+
+  EXPECT_EQ(dict_segment->lower_bound(AllTypeVariant{"Alexander"}), INVALID_VALUE_ID);
+  EXPECT_EQ(dict_segment->upper_bound(AllTypeVariant{"Alexander"}), INVALID_VALUE_ID);
+}
+
+TEST_F(StorageDictionarySegmentTest, CompressSegmentDuplicateValues) {
+  value_segment_int->append(1);
+  value_segment_int->append(1);
+  value_segment_int->append(2);
+  value_segment_int->append(2);
+  value_segment_int->append(1);
+  value_segment_int->append(2);
+
+  const auto dict_segment = std::make_shared<DictionarySegment<int32_t>>(value_segment_int);
+
+  // Test attribute_vector size.
+  EXPECT_EQ(dict_segment->size(), 6);
+
+  // Test dictionary size (uniqueness).
+  EXPECT_EQ(dict_segment->unique_values_count(), 2);
+
+  // Test sorting.
+  EXPECT_EQ(dict_segment->get(0), 1);
+  EXPECT_EQ(dict_segment->get(1), 1);
+  EXPECT_EQ(dict_segment->get(2), 2);
+  EXPECT_EQ(dict_segment->get(3), 2);
+  EXPECT_EQ(dict_segment->get(4), 1);
+}
+
+TEST_F(StorageDictionarySegmentTest, LowerUpperBound) {
+  for (auto value = int16_t{0}; value <= 10; value += 2) {
+    value_segment_int->append(value);
+  }
+
+  std::shared_ptr<AbstractSegment> segment;
+  resolve_data_type("int", [&](auto type) {
+    using Type = typename decltype(type)::type;
+    segment = std::make_shared<DictionarySegment<Type>>(value_segment_int);
+  });
+  auto dict_segment = std::dynamic_pointer_cast<DictionarySegment<int32_t>>(segment);
+
+  EXPECT_EQ(dict_segment->lower_bound(4), ValueID{2});
+  EXPECT_EQ(dict_segment->upper_bound(4), ValueID{3});
+
+  EXPECT_EQ(dict_segment->lower_bound(AllTypeVariant{4}), ValueID{2});
+  EXPECT_EQ(dict_segment->upper_bound(AllTypeVariant{4}), ValueID{3});
+
```

Comment:
```suggestion
EXPECT_THROW(dict_segment->lower_bound(AllTypeVariant{"Opossum"}), std::logic_error);
```
It would be nice to have some test cases testing `lower/upper_bound` with incompatible `AllTypeVariant`s. This would currently fail in your implementation but could be fixed with the suggestion given in `lower_bound`.